### ISSUE TITLE
artefacts in scaled png image with pil engine

### ISSUE THREAD

**Issue by phoet** [python-pillow/Pillow#1449]
i originally created this issue in the thumbor repository, please see that issue for more information: https://github.com/thumbor/thumbor/issues/564

we are seeing some edge-cases where scaled pngs are rendered with artefacts. i'm not sure how i could properly debug this or if i should actually open an issue on the pillow repository. anyhow, maybe someone here can give me a hint on how to proceed.

_those images are white and transparent, you should see nothing on a white background_

this is the original image:

![original](http://image.jimcdn.com/app/cms/image/transf/none/path/s07f3459425dd27f2/image/i23aaf7ddb2a0e16d/version/1443516467/image.png)
http://image.jimcdn.com/app/cms/image/transf/none/path/s07f3459425dd27f2/image/i23aaf7ddb2a0e16d/version/1443516467/image.png

this is what thumbor renderes scaled:

![thumbor scaled](http://image.jimcdn.com/app/cms/image/transf/dimension=259x10000:mode=fitin:format=png/path/s07f3459425dd27f2/image/i23aaf7ddb2a0e16d/version/1443516467/image.png)
http://image.jimcdn.com/app/cms/image/transf/dimension=259x10000:mode=fitin:format=png/path/s07f3459425dd27f2/image/i23aaf7ddb2a0e16d/version/1443516467/image.png

this is what we had before using imagemagic:

![imagemagic scaled](http://u.jimdo.com/www400/o/s07f3459425dd27f2/img/i23aaf7ddb2a0e16d/1443516467/std/image.png)
http://u.jimdo.com/www400/o/s07f3459425dd27f2/img/i23aaf7ddb2a0e16d/1443516467/std/image.png

feedback appreciated

**Comment by wiredfool** [comment]
What's the test script that you're using?   I'm going to guess that it's a loss of precision from going through a premultiplied alpha step to fix artifacts from scaling when there are semitransparent pixels that are scaled.

**Comment by hugovk** [comment]
There's a test script here:
https://github.com/thumbor/thumbor/issues/564#issuecomment-144230835

**Comment by mat** [comment]
This gist contains both @hugovk's script and the original image:

https://gist.github.com/mat/1c18da2d4772c46a6217

@phoet and I are happy for any pointers on how to solve/help debug this.

**Comment by homm** [comment]
You can't fix this with the current implementation of resampling until you are using `NEAREST` method. This problem is same as in `Image.composite()`, which just don't handle transparency right, in opposite to `Image.alpha_composite()`.

**Comment by phoet** [comment]
@homm thx for your feedback. how would using `NEAREST` look like? i honestly have no idea how PIL works internally...

**Comment by homm** [comment]
Nearest neighbor is the cheapest method of resizing when pixels not resampled at all.

**Comment by hugovk** [comment]
@phoet Try something like:
```python
im.thumbnail((259, 96), Image.NEAREST)
```

**Comment by phoet** [comment]
@hugovk @homm thx will take a look into it

**Comment by phoet** [comment]
when creating a thumbnail with `im.thumbnail((259, 95), Image.NEAREST)` it produces an image that does not have black artifacts. allthough from my perception, it has a similarily poor quality compared to what imagemagick generates.

there is quite a difference when comparing those files:
```
compare -metric rmse -compose src image_thumbnail_nearest.png expected_image.png diff.png
8541.14 (0.130329)
```

what would be expected in quality:
![expected_image](https://cloud.githubusercontent.com/assets/48745/10338593/2c2d15ae-6d07-11e5-8546-160d046154cc.png)

what is actually created:
![image_thumbnail_nearest](https://cloud.githubusercontent.com/assets/48745/10338591/2c2783e6-6d07-11e5-8c1e-2014a4999b50.png)

the diff:
![diff](https://cloud.githubusercontent.com/assets/48745/10338592/2c297eb2-6d07-11e5-99ce-bb12020175cc.png)

**Comment by homm** [comment]
This is expected. As I said, nearest is the cheapest method, it can't produce quality result.

**Comment by hugovk** [comment]
So these are the filters:
https://pillow.readthedocs.org/en/3.0.x/handbook/concepts.html#filters

NEAREST is worst. And the default for `thumbnail` is now BICUBIC, but LANCZOS may be better.
https://pillow.readthedocs.org/en/3.0.x/releasenotes/2.7.0.html#default-filter-for-thumbnails

And it may be that Pillow just doesn't have good thumbnailing with transparency.

**Comment by homm** [comment]
> but LANCZOS may be better.

It will produce the same artefacts in this case.

**Comment by phoet** [comment]
@wiredfool i'm curious why you closed this. isn't it still a bug?

**Comment by wiredfool** [comment]
I thought we had come to something of an end point.

**Comment by homm** [comment]
Fixed in #1912

**Comment by phoet** [comment]
oh sweet, will check that out!
